### DISCUSSION TITLE
MPO decomposition into circuits

### DISCUSSION THREAD

**Discussion by ACE07-Sev** [jcmgray/quimb#235] [Q&A]
Greetings Johnnie,

Hope you are doing well good sir. I have been wondering about this for a while now. Given that one can compile a MPS to a quantum circuit to perform state preparation, could one analogously compile a MPO to a quantum circuit to perform unitary (operator) preparation in the same manner?

**Comment by jcmgray** [comment]
Yes sure, one way to think about it is if you just fuse the two legs of each MPO site you get an MPS.

**Reply by ACE07-Sev** [reply]
Greetings dear Johnnie,

Hope you are well sir. Thank you very much for the kind response. Apologies for the delayed response.

 I see. May I expand by asking how expensive it would be? And I apologize if this a stupid question, but would that be as if I am encoding the MPS that is given by contracting the desired MPO with the initial MPS (state) of the circuit in terms of cost?

So instead of directly encoding the circuit representing the MPO, we would be encoding the updated MPS:
![image](https://github.com/jcmgray/quimb/assets/73689800/178b57fd-f519-4ba6-ba00-bbb0e3d3dba5)

So, I am trying to demonstrate quantum computing as a tensor contraction backend, so ideally, I'd like to have the quantum computer perform all of these contractions by directly synthesizing the MPO approximation of a general operator as a circuit and then apply the circuit. That way, the quantum side would be doing the MPO @ MPS = updated MPS part.

Does that make sense? If not, I firstly apologize, and will do my best to clarify.

**Reply by jcmgray** [reply]
I think there is probably a lot of research in this direction, but I am not on top of it. You might be interested in something like https://arxiv.org/pdf/0805.0040 about the cost of contracting some completely general TN via quantum circuit - I presume the further one gets from an isometric form the more costly it gets.

**Reply by ACE07-Sev** [reply]
I see, thank you very much, I'll check it out. I apologize for the confusion though; I meant the cost of fusing the legs of the MPO VS applying the MPO (the MPO we are trying to prepare) to the initial MPS (the state of the circuit before applying the MPO as an MPS) and then encoding the resulting MPS.

These are all done classically, so I mentioned the quantum computer as a contraction backend so that these expensive operations are run on the quantum side instead of the classical one. Again, to clarify one would for instance synthesize a unitary operator to gates directly instead of applying it to the initial state vector classically, and then preparing the resulting state vector. Does that make sense?

I understand the fusing part itself is quite expensive (I may be wrong, hence my first question in this reply), so what is the best way to prepare an MPO? Would it be possible to prepare the MPO directly for instance? Basically, I imagine regardless of making an MPS as you mentioned, or synthesizing an MPO, both would have similar depth (as the depth scaling is with respect to the number of qubits), but with the fusing approach we have this additional step before we prepare the circuit. I'd like to try to synthesize directly so that the MPO is contracted with the initial MPS on the quantum side instead of the classical one.

**Reply by ACE07-Sev** [reply]
Another question I think that may help clarify the above question is that if we have three unitary matrices A, B, and C that we want to apply in that order to a quantum circuit, how should we prepare them when using the MPS/MPO approach?

Should we first contract them together to get a single unitary (and eventually into an MPO) before preparing the circuit, or should we define each as an MPO and prepare them individually? Which would be more efficient?

**Reply by jcmgray** [reply]
Hi @ACE07-Sev, I'm afraid I don't really understand the specifics of your question, and at least personally I only really have time currently to answer specific questions of a technical nature relating to `quimb` here.

**Reply by ACE07-Sev** [reply]
I see, I understand. My apologies for the lengthy question, and thank you very much for the kind guidance thus far. May I conclude the question with how I can do the fusing to get a `qtn.MatrixProductState` instance?
```py
@staticmethod
    def from_operator(operator: NestedCollection[NumberType]) -> qtn.MatrixProductState:
        """ Define the MPS from the operator.

        Parameters
        ----------
        `operator` : NestedCollection[NumberType]
            The operator of the quantum system.

        Returns
        -------
        `mps` : qtn.MatrixProductState
            The MPS representation of the operator.

        Usage
        -----
        >>> operator = np.array([[1, 0], [0, 1]])
        >>> mps = MPS.from_operator(operator)
        """
        # Define the number of sites
        num_sites = int(np.log2(operator.shape[0]))

        # Reshape the operator to N sites where N is the number of qubits
        # needed to represent the operator
        site_dimensions = [2] * num_sites

        # Generate MPO from the tensor arrays
        mpo = qtn.MatrixProductOperator.from_dense(operator, site_dimensions)

        # Fuse the legs of the MPO to construct the MPS
        # TODO

        return mps
```
And if you have the time someday, I'd love to know if there are other ways (which would preferably be more efficient) to prepare the MPO besides fusing.

**Reply by jcmgray** [reply]
Fusing just involves 'reinterpreting' the two local indices/hilbert spaces as joint, so there is not any real cost classically. Here's how you would do it in `quimb`:
```python
import quimb.tensor as qtn

mpo = qtn.MPO_rand(10, 7)

for site in mpo.sites:
    t = mpo[site]
    # the current 'ket' and 'bra' indices
    kix = mpo.upper_ind(site)
    bix = mpo.lower_ind(site)
    # the desired 'physical' index
    pix = f"k{site}"
    # fuse them == reshape == merge
    t.fuse_({pix: [kix, bix]})

# MPS with physical index size 4
mps = mpo.view_as_(qtn.MatrixProductState, site_ind_id="k{}")
```

Whether this translates to the circuit depends on what you want to do with the MPO. E.g. 'prepare' it could mean prepare the vectorized wavefunction, or if it is positive-semi definite, the actual mixed state. Or 'implement' it could mean apply it as a unitary to some existing state.

**Reply by ACE07-Sev** [reply]
So, this is what I'd like to be able to do.

0) Prerequisite: Have a circuit `qc` that you want to apply the operator to
1) Define an arbitrary unitary matrix (operator) $A$
2) Convert the operator $A$ to an MPO `mpo` using `mpo.from_dense(operator, num_qubits)`
3) Fuse the legs of the `mpo` and convert it to a `qtn.MatrixProductState` called `mps`
4) Use the existing encoding from before (I have done MPS to circuit before to do state preparation) to prepare the `mps` as a circuit `op`
5) Add the circuit `op` to the original circuit 

So as I mentioned in the first message in this discussion, I have been using MPS to do state preparation (vectorized wavefunction), and now want to use MPO to do unitary preparation (like qiskit's `qc.unitary(operator, indices)` to approximate an operator and apply it to the circuit. I guess you also mentioned this in your last message above
"Or 'implement' it could mean apply it as a unitary to some existing state."

So, the unitary we approximated using the MPO would be converted to an MPS, and by preparing that MPS we get a circuit `op`. Would it be correct to say by appending `op` to the existing circuit (which would have a state $\ket{\psi}$) we would be essentially doing $A\ket {\psi}$? With this I would then be able to approximate what sth like `qc.unitary()` does but much more efficiently with only a negligible infidelity.

If I can do this by just fusing the MPO into an MPS, then that saves me a great amount of development as I can just use the same code I have that prepares MPS to prepare MPOs as well (since I'd only have to define an MPS from a matrix using (1) to (3), and the rest would be the same).

**Reply by jcmgray** [reply]
So if the task is given a unitary $A$, and $|\psi\rangle$ given by some circuit, get $A|\psi\rangle$, and you have $A$ in dense form, I don't see many advantages to converting it to an MPO first and then a circuit, when you could compile it directly to a set of local gates (I am not familiar with qiskit or what `qc.unitary` does - but probably this?). 

In the case that you are given a unitary in MPO form, then you don't want to fuse the legs instead you want one leg as input and one as output, and to compile each local tensor to a set of local gates on its own:

<img width="850" alt="Screenshot 2024-06-14 at 10 58 50 AM" src="https://github.com/jcmgray/quimb/assets/8982598/27f4126a-4098-4a13-9c78-4e34e18eac94">

where the new inside boxes represent some local optimized circuit structure.

**Reply by ACE07-Sev** [reply]
> So if the task is given a unitary A, and |ψ⟩ given by some circuit, get A|ψ⟩, and you have A in dense form.

Yes, exactly. We're preparing the circuit `op` that represent $A$ so that when we apply it to a circuit `qc` with a state $\ket{\psi}$ we would get $A\ket{\psi}$. For conciseness I'll be referring to this task as $T$.

> I don't see many advantages to converting it to an MPO first and then a circuit, when you could compile it directly to a set of local gates (I am not familiar with qiskit or what `qc.unitary` does - but probably this?).

Well, my hope is that the advantage would be similar to when I did state preparation. By first converting the state to an MPS and then preparing that MPS I was able to get a near exponential reduction in the depth of the circuit (i.e., for a 22-qubit state which was a 1080p image I went from ~8,388,608 depth to 12113 depth which is a 99.85 percent reduction, and I retained a 97 percent fidelity).
![image](https://github.com/jcmgray/quimb/assets/73689800/bdd636fe-e1ef-4d79-bf27-d81683f11d81)
I'm trying to bring the same advantage for this preparation of $A$, hence why we'd have to convert it to an MPO, or with fusing an MPS first. 

> In the case that you are given a unitary in MPO form, then you don't want to fuse the legs instead you want one leg as input and one as output, and to compile each local tensor to a set of local gates on its own, where the new inside boxes represent some local optimized circuit structure.

I'm not sure if I fully understand this. Is this analogous to what I did for state preparation in the context of unitary preparation? Would I be able to get similar depth reduction with the new approach you mentioned above as opposed to the approach below? 

NOTE: By similar depth reduction, I mean going from $O(2^N)$ to $O(N)$.

If we're not doing that (meaning we are doing an exact encoding) that's what [`qc.unitary()`](https://docs.quantum.ibm.com/api/qiskit/qiskit.circuit.library.UnitaryGate) does which scales $O(2^N) $. I am trying to reduce the depth cost like I did with state preparation. That is why I need to define this MPS/MPO representation of $A$ in between the encoding to a circuit. We're basically trying to approximate the unitary using MPS/MPO so that when we keep the bond dimension constant, we can get that exponential depth reduction to $O(N)$.

> So, this is what I'd like to be able to do.
> 
> 0. Prerequisite: Have a circuit `qc` that you want to apply the operator to
> 1. Define an arbitrary unitary matrix (operator) A
> 2. Convert the operator A to an MPO `mpo` using `mpo.from_dense(operator, num_qubits)`
> 3. Fuse the legs of the `mpo` and convert it to a `qtn.MatrixProductState` called `mps`
> 4. Use the existing encoding from before (I have done MPS to circuit before to do state preparation) to prepare the `mps` as a circuit `op`
> 5. Add the circuit `op` to the original circuit
> 
> So as I mentioned in the first message in this discussion, I have been using MPS to do state preparation (vectorized wavefunction), and now want to use MPO to do unitary preparation (like qiskit's `qc.unitary(operator, indices)` to approximate an operator and apply it to the circuit. I guess you also mentioned this in your last message above "Or 'implement' it could mean apply it as a unitary to some existing state."
> 
> So, the unitary we approximated using the MPO would be converted to an MPS, and by preparing that MPS we get a circuit `op`. Would it be correct to say by appending `op` to the existing circuit (which would have a state |ψ⟩) we would be essentially doing A|ψ⟩? With this I would then be able to approximate what sth like `qc.unitary()` does but much more efficiently with only a negligible infidelity.
> 
> If I can do this by just fusing the MPO into an MPS, then that saves me a great amount of development as I can just use the same code I have that prepares MPS to prepare MPOs as well (since I'd only have to define an MPS from a matrix using (1) to (3), and the rest would be the same).

Was I correct about this? Would (1) to (5) perform $T$?

P.S : I sincerely apologize for the numerous edits on this comment. I am trying to make it as brief and clear as possible. I truly hope it doesn't send a notification every time I do.

**Reply by ACE07-Sev** [reply]
Something I just noticed (which I perhaps should have noticed much earlier, apologies) is that if we fuse the physical legs of the MPO at each site it gives an MPS with twice the number of qubits at each site (physical dimension of 4), correct? I hope I haven't misunderstood this.

If so, how would we be representing an operator that acts on N qubits with an MPS for 2N qubits?

**Reply by jcmgray** [reply]
I appreciate you expanding on your questions, but again these are basically theoretical questions that might be better asked e.g. on https://quantumcomputing.stackexchange.com/ or by reviewing the tensor network literature. To briefly mention some considerations:

1. Representing an arbitrary dense unitary and converting it to an MPO scales like O(2^N), so this cost dominates regardless.
2. All operators (up to normalization) can be thought of as pure states in a space twice as large, whether it is useful in this form depends very much on what you want to do with it.

**Reply by ACE07-Sev** [reply]
Understood. Thank you very much for the time and kind guidance!

**Reply by ACE07-Sev** [reply]
Just to provide some concluding thoughts.

> All operators (up to normalization) can be thought of as pure states in a space twice as large, whether it is useful in this form depends very much on what you want to do with it.

What we want to do with this is as before to perform $T$.  With (2) I think it's not really feasible to perform $T$ as the number of qubits wouldn't match.

> Representing an arbitrary dense unitary and converting it to an MPO scales like O(2^N), so this cost dominates regardless.

I don't quite follow this. Why would unitary preparation differ from state preparation in terms of the benefit they'd receive from MPS/MPO approach? Wouldn't it also be O(N)?

I'd deeply appreciate any resources relevant to $T$ using MPOs and how we would prepare the resulting MPO as a circuit. I won't expand further as you requested, and again would like to extend many thanks for the discussion.

**Reply by jcmgray** [reply]
> What we want to do with this is as before to perform T.
 
Here my point was, perform T i.e. prepare or perform A and then do what with it? What form you can have it in will depend on this.

> I don't quite follow this. Why would unitary preparation differ from state preparation in terms of the benefit they'd receive from MPS/MPO approach?

Here I was just making the point that if you are starting by constructing the MPO (or MPS too) by creating it from a dense array then this classical step scales exponentially in N already. It might still be a useful thing to do (depending on the overall task, see above) but if you want a linearly scaling routine then you will need to start with something already in MPS/MPO form.

**Reply by ACE07-Sev** [reply]
> Here my point was, perform T i.e. prepare or perform A and then do what with it? What form you can have it in will depend on this.

I see. So, given the circuit `qc` with N qubits (representing a state $\ket{\psi}$) and an operator $A$ applied to N qubits, we want to prepare the circuit `op` representing $A$ using an MPO. With this, when we append `op` to `qc` we would effectively be doing $A\ket{\psi}$.

I'd like to know how this would be done, and whether it would be similar to the MPS approach (basically being able to prepare the MPO with O(N) circuit depth).

> Here I was just making the point that if you are starting by constructing the MPO (or MPS too) by creating it from a dense array then this classical step scales exponentially in N already.

Ohh I see your point. No no, my statement on $O(N)$ is solely regarding the circuit depth to prepare the state or operator (going from arbitrary MPS/MPO to circuit). I understand constructing MPS/MPO (going from state/operator to MPS/MPO) is classically expensive. 

> It might still be a useful thing to do (depending on the overall task, see above) but if you want a linearly scaling routine then you will need to start with something already in MPS/MPO form.

Right, the motivation is to enable the implementation of deep circuits on current hardware. We achieve this by first approximating them using MPS/MPO, and then subsequently preparing the resulting MPS/MPO to achieve $O(N)$ circuit depth as opposed to $O(2^N)$.

For example, with the state preparation we had to construct the MPS from the state vector (using `mps.from_dense()`) which is classically expensive. However, with this MPS we were able to subsequently prepare a circuit which would not have been feasible on any current hardware before (where we went from 8 million depth to 12 thousand, a massive reduction in the circuit depth).

So, the goal is to compress these deep circuits using this approximate preparation approach. It would allow us to answer "can vs can't". Right now, we can't do any practical state preparation/unitary preparation for states with more than 10 qubits or so. The circuits become too deep. With the MPS/MPO approach we will actually be able to do these, but of course the gotcha will be the classical cost to construct the MPS/MPO. When better alternatives present themselves, we can then evaluate "should vs shouldn't".
